### PR TITLE
feat: produce phase0 beacon block body once

### LINKED ISSUE #6159: Produce phase0 beacon block body once
### Problem description

Right now if both builder and execution block are configured, we produce beacon block body twice. 

### Solution description

We should be able to do it once to speed up the publish process. Also look into other steps to see if we can publish block faster to reduce our reorg percentage https://reorg.pics/

cc @nazarhussain @g11tech 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

Make the block production more efficient. 

**Description**

- Produce the beacon-block-phase0 once during the race of execution and builder blocks. 

Closes #6159 

**Steps to test or reproduce**

- Run all tests.

### CODE REVIEW

File: packages/beacon-node/src/chain/chain.ts
Comment (nazarhussain) [inline packages/beacon-node/src/chain/chain.ts:485]: To avoid changes in the metrics structure for now in this PR I had to fix this type for now. We may think to refactor the metrics structure after merging this PR.

File: packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts
Comment (g11tech) [inline packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts]:
```suggestion
    commonBlockBody,
```
lets not call it phase0 block body since there can be common fields like (blsToExecution) and its type can be Omit<allForks.BeaconBlockBody,"executionPayload">

File: packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts
Comment (g11tech) [inline packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts]: move this also into produceCommonBlockBody, so whatever which isn't dependent on execution payload/header, which is almost all fields for now since all other fields are independent

File: packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts
Comment (g11tech) [inline packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts]:
```suggestion
export async function produceBlockBodyCommon<T extends BlockType>(
```

File: packages/beacon-node/src/chain/interface.ts
Comment (nazarhussain) [inline packages/beacon-node/src/chain/interface.ts:213]: `Omit<allForks.BeaconBlockBody, "executionPayload">` is not type-safe, so why used explicit attributes here.

**Review by g11tech — APPROVED** [review]
lgtm

**Review by g11tech — CHANGES_REQUESTED** [review]
need to see why sim merge are failing else looks good (retriggered the run as it failed all image pull for geth)
also unit tests are failing @nazarhussain

**Review by g11tech — DISMISSED** [review]
approving, sim merge seem to be failing because of docker pull limit, retry in couple of hours and merge in if everytthing passes @nazarhussain

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (twoeths) [inline packages/beacon-node/src/api/impl/validator/index.ts:291]: I don't see the usage of this type `Omit<routes.validator.ExtraProduceBlockOps, "builderSelection">`, can we drop it?

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (twoeths) [inline packages/beacon-node/src/api/impl/validator/index.ts:293]: why do we type this as optional while we always provide value and we have to handle fallback logic downstream?

File: packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts
Comment (nflaig) [inline packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:589]: Is it still required to have 2 different metrics for this? maybe should be consolidated into a single metric

There is also the option now to return blinded local blocks (https://github.com/ChainSafe/lodestar/pull/6227), metric naming should reflect that.

File: packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts
Comment (nflaig) [inline packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:573]: Why do we need to pass blockType here? Shouldn't be relevant for this function

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (nflaig) [inline packages/beacon-node/src/api/impl/validator/index.ts]: `produceCommonBlockBody` does not only produce phase 0 blocks

File: packages/beacon-node/src/chain/chain.ts
Comment (nflaig) [reply packages/beacon-node/src/chain/chain.ts:485]: Shouldn't metric consolidation be part of this PR, feel like passing blockType around just for metrics is not ideal

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (nazarhussain) [reply packages/beacon-node/src/api/impl/validator/index.ts:293]: We don't always provide option, one can directly call `chain.produceBlock` and it will still be able to produce the block. We are doing it in our code for `produceBlockV2`

File: packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts
Comment (nazarhussain) [reply packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:589]: Mentioned it earlier in discussion that this metric need cleanup.

File: packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts
Comment (nazarhussain) [reply packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:573]: So we can pass call this for either type of block.

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (nazarhussain) [reply packages/beacon-node/src/api/impl/validator/index.ts]: It was producing only phase0 earlier.

File: packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts
Comment (nflaig) [reply packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:573]: But why do we need to know the block type at all? The function produces the same result irrespective of the block type we pass here (or at least that's what I understood is the goal of this PR, to consolidate body production and only do it once).

Comment (nazarhussain) [reply packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:573]: As of current implementation we need to know block type to update respective metric. Once we consolidate metrics then we don't need block type for common block body.

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (nazarhussain) [reply packages/beacon-node/src/api/impl/validator/index.ts:291]: Somehow it's used on the line 400 of this file. I will clean this up in a separate PR in-progress to clean up this flow.

**Review by g11tech — APPROVED** [review]
as per discussions with @nazarhussain and other comments and my own opinions on how metrics need to be included for the common (my view point: just duplicate them on full and blinded panels whenever body build is triggered as a common path), seems like we need discussion on this so that we have consensus on what is useful for the team.

approving now with an understanding for an immediate metrics consolidation followup PR to be done by @nazarhussain post discussion on how reflect the metrics of common path

File: packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts
Comment (nflaig) [reply packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts:573]: Let's keep metrics as is for now then, will give this another review locally as it is probably easier that way

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.14.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.14.0) :tada:

**Pull request merged**
